From 2dbbded3a39cbb5b91a36a01a0fcd3ac1c360ade Mon Sep 17 00:00:00 2001
From: Naveen Mamindlapalli <naveenm@marvell.com>
Date: Fri, 21 Aug 2020 15:36:47 +0530
Subject: [PATCH 625/767] octeontx2-af: Fix updating wrong multicast list index
 in NIX_RX_ACTION

commit 0dcc41e27a336144457985e8803d7abe6dd76b05 from
git@git.assembla.com:cavium/WindRiver.linux.git

This patch fixes the issue of updating wrong multicast list index
in NIX_RX_ACTION when VF comes up. Otherwise, the NIX hw block
receiving the broadcast traffic on the PF will attempt to refer
to the incorrect multicast list index entry NIX_RX_MCE_S which was
initialized to direct the traffic to another PF FUNC that may not
have been initialized resulting in SMMU translation faults.

Fixes: 1f2c72a130a ("octeontx2-af: Add NIX1 interfaces to NPC")
Change-Id: Iad13241c7a78c40ae6bffa1be3a4c127222096de
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/34369
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c
index 5e05b6490c9f..01277fc1d938 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c
@@ -719,11 +719,11 @@ void rvu_npc_enable_promisc_entry(struct rvu *rvu, u16 pcifunc, int nixlf)
 void rvu_npc_install_bcast_match_entry(struct rvu *rvu, u16 pcifunc,
 				       int nixlf, u64 chan)
 {
-	struct rvu_pfvf *pfvf = rvu_get_pfvf(rvu, pcifunc);
 	struct npc_mcam *mcam = &rvu->hw->mcam;
 	struct mcam_entry entry = { {0} };
 	struct rvu_hwinfo *hw = rvu->hw;
 	struct nix_rx_action action;
+	struct rvu_pfvf *pfvf;
 	int blkaddr, index;
 
 	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_NPC, 0);
@@ -742,6 +742,7 @@ void rvu_npc_install_bcast_match_entry(struct rvu *rvu, u16 pcifunc,
 
 	/* Get 'pcifunc' of PF device */
 	pcifunc = pcifunc & ~RVU_PFVF_FUNC_MASK;
+	pfvf = rvu_get_pfvf(rvu, pcifunc);
 	index = npc_get_nixlf_mcam_index(mcam, pcifunc,
 					 nixlf, NIXLF_BCAST_ENTRY);
 
-- 
2.31.1

